### PR TITLE
[Proposal] Automated performance tests for brew

### PR DESCRIPTION
**Pull request by zachauten**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -106,6 +106,16 @@ As part of the installation process, Homebrew rewrites its compiled binaries to
 Expected outcome: A refactoring of Homebrew's Linux-specific `keg` and `keg_relocate` interfaces to use `rbelftools` and `patchelf.rb` instead of shelling out to the GNU utilities.
 
 ----
+### Automated Performance tests for brew
+#### Skills Required: Ruby, Homebrew usage, Performance Testing, Tracing tools.
+#### Mentor: @zachauten
+#### Difficulty: Medium
+
+Write a suite of automated performance tests that can run against brew’s code base to generate benchmarks, and identify performance bottlenecks. These tests could also be added to brew’s CI using Github Actions, o detect regressions in performance when new code is committed.
```
Comment (MikeMcQuaid) [inline README.md]:
```suggestion
Write a suite of automated performance tests that can run against brew’s code base to generate benchmarks, and identify performance bottlenecks. These tests could also be added to brew’s CI using GitHub Actions, o detect regressions in performance when new code is committed.
```

File: README.md [hunk 2]
```diff
@@ -106,6 +106,16 @@ As part of the installation process, Homebrew rewrites its compiled binaries to
 Expected outcome: A refactoring of Homebrew's Linux-specific `keg` and `keg_relocate` interfaces to use `rbelftools` and `patchelf.rb` instead of shelling out to the GNU utilities.
 
 ----
+### Automated Performance tests for brew
+#### Skills Required: Ruby, Homebrew usage, Performance Testing, Tracing tools.
```
Comment (MikeMcQuaid) [inline README.md]: Would be good to list specific tools here. `brew prof` provides tracing but we should recommend what the performance testing tools would look like if this is to be included.

Comment (zachauten) [reply README.md]: Added!

**Comment by MikeMcQuaid** [comment]
@zachauten Sorry, I'm still not sure exactly what's being proposed here or how it would be implemented. Would this be using `rspec` or something else? Would it be run on every commit? If so: I'm concerned it may add a lot to CI time. If not: I'm concerned it will not get noticed.

**Comment by zachauten** [comment]
@MikeMcQuaid Yes I think rspec would be appropriate since it's what's being used for unit tests.

I think after creating a performance testing suite, adding it to CI is the logical next step because it would need to be run on regular intervals to be useful. I can understand the concern about the time it would add to CI. Maybe it could run on commits but not prevent merges, as a sort of canary? Or run after a certain number of commits.

I know the submission deadline for gsoc is soon, so if you think this is something that needs much more discussion or may not be useful perhaps it would be better to create an issue in brew than a proposal here?

**Comment by MikeMcQuaid** [comment]
> I know the submission deadline for gsoc is soon, so if you think this is something that needs much more discussion or may not be useful perhaps it would be better to create an issue in brew than a proposal here?

Yeh, I think so. I think this is in the camp of "I'd need to review a PR rather than reviewing an issue or proposal". If you're game to make one I'd love to see it!

**Pull request closed without merging**